File: main/fh.m
function [output, gradient] =  fh(chol_matrix, sigma_inv, S, mu, weight)
% FH provides output and gradient of objective function for covariance
% update given the example selected by human. It is important to provide
% the covariance to accelerate computation and not to overflow the memory.
%
% Input arguments:
%   chol_matrix     - Cholesky decomposition of covariance matrix of
%                     variational distribution
%   sigma_inv       - Inverse of prior covariance matrix.
%   S               - Embeddings of words in set.
%   mu              - Mean of variational distribution.
%   weight          - Scale regularization to prior.
%
% Returns:
%   output     - Value of objective function for given variational matrix.
%   gradient   - Gradient of objective function at given variational matrix.

    %  If weight is not given, set to 1
    if nargin < 5
        weight = 1;
    end

    % Ensuring that the Cholesky decomposition, cholMatrix, is 
    % a lower triangular matrix with real and positive diagonal entries.
    chol_matrix = tril(chol_matrix);
    neg_daig_ind = find(diag(chol_matrix)<=0);
    chol_matrix(neg_daig_ind, neg_daig_ind) = 10^(-14); 
    
    % Compute the output
    exp_term = bsxfun(@plus, S' * mu, 0.5 * sum((S' * (chol_matrix*chol_matrix')) .* S', 2));
    log_sum = log(sum(exp(exp_term), 1));
    output = -weight*sum(log(diag(chol_matrix))) + weight*0.5 * trace(sigma_inv*(chol_matrix*chol_matrix'))...
        + log_sum;
    
    % Compute gradient
    LLt = chol_matrix * chol_matrix';
    quadratic_terms = sum((S' * LLt) .* S', 2); % S' * (L*L') * S for each column
    exp_terms = exp(S' * mu + 0.5 * quadratic_terms);

    gradient = weight * diag(1 ./ diag(chol_matrix)) ...
               - ((weight * sigma_inv) ...
               + S * diag(exp_terms) * S') * chol_matrix;
    gradient = -gradient;    
 
    % Check for numerical stability: Adjust if gradient norm is too large
    if norm(gradient)> 1e30
        gradient = gradient ./ 1e10;
    end
 
    % Display warnings in case of NaN or Inf in gradient or output
    if isnan(gradient)
        disp("Gradient has NaN values");
    elseif isinf(gradient)
        disp("Gradient has Inf values");
    end
    if isnan(output)
        disp("Objective function output has NaN values");
    elseif isinf(output)
        disp("Objective function output has Inf values");
    end
end

File: python/fh.py
import numpy as np

# If weight is not given, set it to 1
def fh(chol_matrix_flat, sigma_inv, S, mu, weight = 1, shape = (3, 3)):
    """
    Provides output and gradient of the objective function for covariance update 
    given the example selected by a human.
    
    Parameters:
    chol_matrix    - Cholesky decomposition of  the covariance matrix of 
                     the variational distribution (lower triangular matrix).
    sigma_inv      - Inverse of the prior covariance matrix.
    S              - Embeddings of words in the set.
    mu             - Mean of the variational distribution.
    weight         - Scale regularization to prior (default: 1).
    
    Returns:
    output         - Value of the objective function.
    gradient       - Gradient of the objective function at the given variational matrix.
    """
    chol_matrix = chol_matrix_flat.reshape(shape)
    # Ensuring that the Cholesky decomposition is a lower triangular matrix
    chol_matrix = np.tril(chol_matrix) # OK
    # Ensure no negative diagonal entries
    diag_values = np.diag(chol_matrix) # OK
    neg_diag_ind = np.where(diag_values <= 0)[0] # OK
    chol_matrix[neg_diag_ind, neg_diag_ind] = 10 ** (-14) # OK
    # Compute exp_term and log_sum
    exp_term = (S.T @ mu.T).reshape(-1, 1) + (0.5 * np.sum((S.T @ (chol_matrix @ chol_matrix.T)) * S.T, axis=1)).reshape(-1, 1) # OK
    log_sum = np.log(np.sum(np.exp(exp_term), axis=0)) # OK
    # Compute the final output
    output = -weight * np.sum(np.log(np.diag(chol_matrix))) + weight * 0.5 * np.trace(sigma_inv @ (chol_matrix @ chol_matrix.T)) + log_sum # OK
    # Compute the gradient (LLt and quadratic terms)
    LLt = chol_matrix @ chol_matrix.T # OK
    quadratic_terms = np.sum((S.T @ LLt) * S.T, axis=1).reshape(-1, 1)  # OK
    exp_terms = np.exp((S.T @ mu.T).reshape(-1, 1) + 0.5 * quadratic_terms) # OK
    exp_terms_flat = exp_terms.flatten() # OK
    # Compute the gradient
    gradient = weight * np.diag(1.0 / np.diag(chol_matrix)) \
           - ((weight * sigma_inv) + S @ np.diag(exp_terms_flat) @ S.T) @ chol_matrix
    gradient = -gradient # OK
    # Check for numerical stability: Adjust if gradient norm is too large
    if np.linalg.norm(gradient) > 1e30:
        gradient = gradient / 1e10
    # OK
    # Display warnings in case of NaN or Inf in gradient or output
    if np.isnan(gradient).any():
        print("Gradient has NaN values")
    elif np.isinf(gradient).any():
        print("Gradient has Inf values")
    
    if np.isnan(output):
        print("Objective function output has NaN values")
    elif np.isinf(output):
        print("Objective function output has Inf values")

    return output, gradient.flatten()
    #return output
